sdk/javascript/src/RuleBasedTransactionFactory.js

```diff
@@ -0,0 +1,195 @@
+		else if (0 === hint.indexOf('struct:'))
+			ruleName = hint;
+
+		if (ruleName)
+			typeHints[key] = ruleName;
+	});
+
+	return typeHints;
+};
+
+const typeConverterFactory = (module, customTypeConverter, value) => {
+	if (customTypeConverter && customTypeConverter(value))
+		return customTypeConverter(value);
+
+	if (value instanceof ByteArray)
+		return new module[value.constructor.name](value.bytes);
+
+	return value;
+};
+
+const autoEncodeStrings = entity => {
+	Object.getOwnPropertyNames(entity).forEach(key => {
+		const value = entity[key];
+		if ('string' === typeof (value))
+			entity[key] = new Uint8Array(Array.from(value).map(ch => ch.charCodeAt(0)));
+	});
+};
+
+/**
+ * Rule based transaction factory.
+ */
+class RuleBasedTransactionFactory {
+	/**
+	 * Creates a rule based transaction factory for use with catbuffer generated code.
+	 * @param {object} module Catbuffer generated module.
+	 * @param {function} typeConverter Type converter.
+	 * @param {Map} typeRuleOverrides Type rule overrides.
+	 */
+	constructor(module, typeConverter = undefined, typeRuleOverrides = undefined) {
+		this.module = module;
+		this.typeConverter = value => typeConverterFactory(this.module, typeConverter, value);
+		this.typeRuleOverrides = typeRuleOverrides || new Map();
+		this.rules = new Map();
+	}
+
+	_getModuleClass(name) {
+		return this.module[name];
+	}
+
+	/**
+	 * Creates wrapper for SDK POD types.
+	 * @param {string} name Class name.
+	 * @param {type} PodClass Class type.
+	 */
+	addPodParser(name, PodClass) {
+		if (this.typeRuleOverrides.has(PodClass)) {
+			this.rules.set(name, this.typeRuleOverrides.get(PodClass));
+			return;
+		}
+
+		this.rules.set(name, value => (value instanceof PodClass ? value : new PodClass(value)));
+	}
+
+	/**
+	 * Creates flag type parser.
+	 * @param {string} name Class name.
+	 */
+	addFlagsParser(name) {
+		const FlagsClass = this._getModuleClass(name);
+		const stringToEnum = buildEnumStringToValueMap(FlagsClass);
+
+		this.rules.set(name, flags => {
+			if ('string' === typeof (flags)) {
+				const enumArray = flags.split(' ').map(flagName => nameToEnumValue(stringToEnum, name, flagName));
+				return FlagsClass.or(...enumArray);
+			}
+
+			if ('number' === typeof (flags))
+				return new FlagsClass(flags);
+
+			return flags;
+		});
+	}
+
+	/**
+	 * Creates enum type parser.
+	 * @param {string} name Class name.
+	 */
+	addEnumParser(name) {
+		const EnumClass = this._getModuleClass(name);
+		const stringToEnum = buildEnumStringToValueMap(EnumClass);
+
+		this.rules.set(name, enumValue => {
+			if ('string' === typeof (enumValue))
+				return nameToEnumValue(stringToEnum, name, enumValue);
+
+			if ('number' === typeof (enumValue))
+				return new EnumClass(enumValue);
+
+			return enumValue;
+		});
+	}
+
+	/**
+	 * Creates struct parser (to allow nested parsing).
+	 * @param {string} name Class name.
+	 */
+	addStructParser(name) {
+		const StructClass = this._getModuleClass(name);
```

Comment:
Not sure if lint allows that, but wondering if we should use names like Struct_Class or StructClass in py version as well (instead of current struct_class)